### PR TITLE
[x264] Update and fixes

### LINKED ISSUE #22899: [x264] arm64 cross-compile settings are ignored.
**Describe the bug**
[x264] port does **not** propagate cross-compile settings to the 'configure' script. The package is built using host settings leading to output executable format mismatch (if expected output is aarch64 and host is x86_64, for ex).

**Host Environment**
- OS: Linux
- Arch: x86_64
- Compiler: gcc 10.3.0
- Cross-Compiler: linaro arm64 cross-compiler v7.5.0

**Target Environment**
- OS: Linux
- Arch: arm64
- Compiler: gcc linaro v7.5.0

**To Reproduce**

1. Use triplet with any cross-compile toolchain (via VCPKG_CHAINLOAD_TOOLCHAIN_FILE)
2. Add vcpkg-cmake to vcpkg.json
3. Add following lines to portfile
```
vcpkg_cmake_get_vars(cmake_vars_file)
include("${cmake_vars_file}")
```
4. Build
5. Verify that VCPKG_DETECTED_CMAKE_CROSSCOMPILING is set to TRUE in cmake-vars log file.
6. Verify in build log that cross-compile toolchain compiler was **NOT** used.

**Expected behavior**
Build should work, but use the appropriate cross-compile compiler as set per the toolchain.

**Failure logs** 
n.a

**Additional context**
The x264 'configure' options '--host' and '--cross-prefix' are not propagated, hence the build does not use the appropriate tools.

### PR DESCRIPTION
**Pull request by dg0yt**
- #### What does your PR fix?
  Portfile cleanup.
  Fixes a `mkdir` race condition on install, causing occasional CI errors.
  Updates to the latest version/commit used in [upstream binary releases](https://artifacts.videolan.org/x264/).
  Switches from `version-string` to the ordered `version` scheme with the numbers used in pkg-config files (when build from git, or with extra fixup in this port now). There is no other official scheme. Repology finds various variants.
  Makes `asm` (assembly optimization) an explicit feature. 
  Doesn't require nasm when not building for x64 or x86.
  Uses updated gas-preprocessor, which enables arm64-windows builds with assembler optimization.
  Fixes cross-builds by passing actual `CC` and cross prefix. Resolves #22899.
  Makes the tool an explicit port feature.
  Fixes clang-cl build (authored by @Neumann-A).

- #### Which triplets are supported/not supported? Have you updated the [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt)?
  all (new: arm64-windows), no

- #### Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?
  yes

- #### If you have added/updated a port: Have you run `./vcpkg x-add-version --all` and committed the result?
  yes

### CODE REVIEW

**Comment by Neumann-A** [comment]
You might want to add
https://github.com/microsoft/vcpkg/pull/25897/files#diff-9d0f8dd1dc451d1a25a507a79bb90353e1b4fbb7347d91c998460c56c1579663

**Comment by dg0yt** [comment]
@Neumann-A I cherry-picked the clang-cl patch.

**Comment by BillyONeal** [comment]
Excellent, finally an answer on what was causing that funny macos CI failure. Thanks!

File: ports/x264/allow-clang-cl.patch [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+@@ -606,9 +606,9 @@ if [[ $host_os = mingw* || $host_os = msys* || $host_os = cygwin* ]]; then
+         if cc_check '' -Qdiag-error:10006,10157 ; then
+             CHECK_CFLAGS="$CHECK_CFLAGS -Qdiag-error:10006,10157"
+         fi
+-    elif [[ "$cc_base" = cl || "$cc_base" = cl[\ .]* ]]; then
++    elif [[ "$cc_base" = cl || "$cc_base" = cl[\ .]* || "$cc_base" = clang-cl || "$cc_base" = clang-cl[\ .]* ]]; then
+         # Standard Microsoft Visual Studio
+         compiler=CL
+         compiler_style=MS
++        CFLAGS="$CPPFLAGS $CFLAGS -nologo -GS- -DHAVE_STRING_H -I\$(SRCPATH)/extras"
```
Comment (BillyONeal) [inline ports/x264/allow-clang-cl.patch:14]: `-GS-` is a bit yikes but it's in the upstream I guess :/

No change requested.

File: ports/x264/portfile.cmake [hunk 2]
```diff
@@ -1,94 +1,139 @@
...
+endif()
 
-vcpkg_find_acquire_program(NASM)
-get_filename_component(NASM_EXE_PATH ${NASM} DIRECTORY)
-vcpkg_add_to_path(${NASM_EXE_PATH})
+vcpkg_list(SET EXTRA_ARGS)
+set(nasm_archs x86 x64)
+set(gaspp_archs arm arm64)
+if(NOT "asm" IN_LIST FEATURES)
+    vcpkg_list(APPEND OPTIONS --disable-asm)
```
Comment (BillyONeal) [inline ports/x264/portfile.cmake:55]: Why would someone ever want this feature to be off? It seems like we should just set this to (morally) `x86 | x64 | (arm & !windows) | arm64`

**Comment by JackBoosY** [comment]
Okay, can you please resolve the file conflicts?

Thanks.

**Review by BillyONeal — COMMENTED** [review]
This rocks! Just curious about that asm feature...

**Comment by BillyONeal** [comment]
I resolved the conflict with https://github.com/microsoft/vcpkg/pull/26371 by just ignoring all changes from there in favor of the changes here (you can see that the SHA in the version database didn't change) because the effect of that change was already here.

**Comment by JackBoosY** [comment]
Looks good now, just waiting for reply.

File: ports/x264/portfile.cmake (hunk 2 above)
Comment (dg0yt) [reply ports/x264/portfile.cmake:55]: I expect that this feature simply won't work for more exotic platforms, so I want to offer control. An alternative control would be a triplet variable.

**Comment by JavierMatosD** [comment]
Awesome! Thanks!

**Comment by BillyONeal** [comment]
Thanks again @dg0yt!

**Pull request merged**